### ISSUE TITLE
[analyzer/ffi] Add tests executing ffi_verifier.dart

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#44773]
We should add tests that exercise `ffi_verifier.dart` during various actions the `dartanalyzer` can do besides analyzing.

* code completion
* hover

That way we might prevent issues such as https://github.com/dart-lang/sdk/issues/44763 in the future.

(How about fuzzing?)

@srawlins @scheglov Currently we only have analyzer tests for warnings/errors that are ffi-specific. Where do I add tests for code completion and hover? And do we have non-ffi analyzer tests for these actions that I can take inspiration from?

For reference:

* pkg/analyzer/lib/src/generated/ffi_verifier.dart - ffi specific implementation
* pkg/analyzer/test/src/diagnostics/(.*)_test.dart - error/warning test paths

There are test called `completion`/`complete`, but they are outside pkg/analyzer.

**Comment by scheglov** [comment]
It is not code completion and hover per se that should be tested. We see exceptions there because it is the nature of Dart Analysis Server to work with incomplete and invalid code - some identifiers might be unresolved because the user has not finished typing them yet. So, the goal here should be to verify that FFI tests have necessary invalid code cases that cover places where something might be unresolved.

**Comment by Yusufihsangorgel** [comment]
Following the ffi_verifier change in https://dart-review.googlesource.com/c/sdk/+/525581 I can take this one next. Plan per the earlier comments: tests/ffi/static_checks cases with invalid and unresolved code so ffi_verifier is exercised where resolution may have failed, plus analyzer unit tests alongside the existing diagnostics tests. Anything you would want covered first?

**Comment by dcharkes** [comment]
> Anything you would want covered first?

Probably to try invalid types and unresolved references in all positions that are used for FFI custom type checking. No idea which corner cases exactly to cover. (Maybe a good job for an LLM: analyze the the ffi transformers in the CFE and the ffi_verifier and come up with Dart programs that might make them crash.)

**Comment by Yusufihsangorgel** [comment]
I'll enumerate the positions from ffi_verifier and the CFE transformer and write a case per position with both an invalid type and an unresolved reference, so the coverage is systematic rather than hand-picked. Will start after the mixin CL lands.

**Comment by Yusufihsangorgel** [comment]
Started going through the FFI verifier positions for this. The first invalid-code case that does more than report a diagnostic is a crash, so flagging it before I go further.

This input crashes the analyzer (analysis server internal error) instead of reporting a diagnostic:

```dart
import 'dart:ffi';
int _cb(int x) => x;
void main() {
  NativeCallable<Int32 Function(Int32)>.isolateLocal(_cb, 0);
}
```

```
Internal error: type 'IntegerLiteralImpl' is not a subtype of type 'NamedArgument' in type cast
```

The cause is in `_validateNativeCallable`. When the return type needs an exceptional return and `argCount == 2`, the second argument is read as `(arguments2[1] as NamedArgument).argumentExpression2`. That assumes the second argument is the named `exceptionalReturn:`, but a positional second argument (the `0` above) is an `IntegerLiteralImpl`, so the cast throws.

The parallel `_validateFromFunction` runs the same exceptional-return check but reads `arguments2[1].argumentExpression` with no cast, which is safe there because `Pointer.fromFunction`'s `exceptionalReturn` is positional. `Pointer.fromFunction<Int32 Function(Int32)>(_cb, 0)` analyzes cleanly. On `NativeCallable.isolateLocal` the slot is a named parameter, so passing it positionally is already invalid, and `_validateNativeCallable` casts to `NamedArgument` without first checking the second argument is actually named.

Before I turn this into a fix plus a static-error test, one question on what you would want here. Passing a positional argument to the named `exceptionalReturn:` is already reported as an extra positional argument, so one option is to guard with `is NamedArgument` and bail without an FFI diagnostic, since the code is already flagged. The other option is to keep an FFI message, reusing `missingExceptionValue` for a non-named second argument. I lean toward bailing so we do not double-report, but it is your call.

I will keep walking the rest of the positions the same way and collect anything else that crashes rather than diagnoses.

**Comment by bwilkerson** [comment]
> ... bail without an FFI diagnostic, since the code is already flagged.

Yes, please. There's no value in reporting the second "problem" because there's really only one problem here. (And there's a fix that will add the name to the second argument.)

**Comment by Yusufihsangorgel** [comment]
> Yes, please.

That landed as [`ab364ca4`](https://github.com/dart-lang/sdk/commit/ab364ca4aa5a9cef2bfb411c5508924606506e3d) (CL 526980), reviewed by @dcharkes and @johnniwinther. `_validateNativeCallable` walks the argument list once now, picking up the positional callback and the named exceptional return in the same pass, and returns without a diagnostic when it finds a second positional argument. The same restructuring covers named-first calls: `isolateLocal(exceptionalReturn: 4, f)` used to be checked with the exceptional value in the callback's slot. Eight new tests in `ffi_async_callback_test.dart`, one of them the `IntegerLiteralImpl` crash from above. After the bail, `isolateLocal(f, 0)` reports only `extraPositionalArgumentsCouldBeNamed`, and `ConvertToNamedArguments` is already registered for that, so the fix you mentioned comes with the diagnostic that survives.

I've inventoried the 29 `_validate*` and three `_checkFfiNative*` methods in `ffi_verifier`, grouped by which unchecked read each one does. That inventory is local, and the set of invalid and unresolved cases @dcharkes asked for isn't written yet. One method is done, and in that one only the unchecked cast. Its `typeOrThrow` and `typeArguments` reads are still unguarded.

Narrowing what I said on July 20: I'm working the verifier, not the CFE transformer. The transformer runs on kernel the front end already accepted, while the verifier runs on half-resolved IDE ASTs, which is why the #44763 shape reads as analyzer-only to me.

Next is `_validateFromFunction`. I'll post here as each one goes up.

**Comment by Yusufihsangorgel** [comment]
`_validateFromFunction` is up: https://github.com/dart-lang/sdk/pull/63943. All three argument positions take invalid and unresolved input without crashing on current main. A bare import prefix resolves to `InvalidType` rather than a null static type, so both `typeOrThrow` reads hold up. A wrong explicit type-argument count keeps `typeArgumentTypes` populated. So this one is tests only, pinning that. Eight of the pinned FFI messages sit on code another diagnostic already flags, the shape `isolateLocal` still reports too; the PR says why that would be its own change. Next is `_validateAsFunction`.
